libs/contrib/Search/Generator.idr

```diff
@@ -0,0 +1,87 @@
+||| The content of this module is based on the paper
+||| Applications of Applicative Proof Search
+||| by Liam O'Connor
+|||
+||| The main difference is that we use `Colist1` for the type of
+||| generators rather than `Stream`. This allows us to avoid generating
+||| many duplicates when dealing with finite types which are common in
+||| programming (Bool) but even more so in dependently typed programming
+||| (Vect 0, Fin (S n), etc.).
+
+module Search.Generator
+
+import Data.Colist
+import Data.Colist1
+import Data.Fin
+import Data.List
+import Data.Stream
+import Data.Vect
+
+------------------------------------------------------------------------
+-- Interface
+
+||| A generator for a given type is a non-empty colist of values of that
+||| type.
+public export
+interface Generator a where
+  generate : Colist1 a
+
+------------------------------------------------------------------------
+-- Implementations
+
+-- Finite types
+
+||| ALL of the natural numbers
+public export
+Generator Nat where
+  generate = fromStream nats
+
+||| ALL of the booleans
+public export
+Generator Bool where
+  generate = True ::: [False]
+
+||| ALL of the Fins
+public export
+{n : Nat} -> Generator (Fin (S n)) where
+  generate = fromList1 (allFins n)
+
+-- Polymorphic generators
+
+||| We typically want to generate a generator for a unit-terminated right-nested
+||| product so we have this special case that keeps the generator minimal.
+public export
+Generator a => Generator (a, ()) where
+  generate = map (,()) generate
+
+||| Put two generators together by exploring the plane they define.
+||| This uses Cantor's zig zag traversal
+public export
+(Generator a, Generator b) => Generator (a, b) where
+  generate = plane generate (\ _ => generate)
+
+||| Put two generators together by exploring the plane they define.
+||| This uses Cantor's zig zag traversal
+public export
+{0 b : a -> Type} -> (Generator a, (x : a) -> Generator (b x)) =>
```

Comment:
Shouldn't the second auto parameter be `{x : a} -> Generator (b x)` instead of `(x : a) -> Generator (b x)`? I mean, even your `Generator (Vect n a)` takes `n` as implicit parameter.

Reply:
> even your Generator (Vect n a) takes n as implicit parameter.

That's down to the parser not allowing anything else. I don't have strong opinions one
way or the other: Idris is writing the code putting these together, not me.

Reply:
I've updated my comment seemingly after you've answered, so I'll repeat: once I had a problem the compiler feeding `((x : a) -> X x) => ...` parameter when there was no problem with alternative one `({x : a} -> X x) => ...`